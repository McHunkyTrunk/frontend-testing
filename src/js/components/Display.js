import React from "react";

const Display = ({mainValue, secondaryValue}) => {
  return (
    <div className="Display" data-testid="Display">
      <span className="main-value" data-testid="Display__value">{mainValue}</span>
      <span className="secondary-value">{secondaryValue}</span>
    </div>
  );
};

export default Display;
