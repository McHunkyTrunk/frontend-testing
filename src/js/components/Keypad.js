import React from "react";

const Keypad = ({onKey}) => {
  const renderNumber = digit => (
    <td
      onClick={() => onKey(digit)}
      className={`digit digit-${digit}`}
      data-digit={digit}
      data-testid={`Keypad__digit-${digit}`}
    >
      {digit}
    </td>
  );

  const renderOperator = (operator, name) => (
    <td
      onClick={() => onKey(operator)}
      className={`operator operator-${name}`}
      data-operator={operator}
      data-testid={`Keypad__operator-${name}`}
    >
      {operator}
    </td>
  );

  return (
    <table className="Keypad" data-testid="Keypad">
      <tbody>
        <tr>
          {renderNumber("7")}
          {renderNumber("8")}
          {renderNumber("9")}
          {renderOperator("+", "plus")}
        </tr>

        <tr>
          {renderNumber("4")}
          {renderNumber("5")}
          {renderNumber("6")}
          {renderOperator("-", "subtract")}
        </tr>

        <tr>
          {renderNumber("1")}
          {renderNumber("2")}
          {renderNumber("3")}
          {renderOperator("*", "multiply")}
        </tr>

        <tr>
          {renderOperator("C", "reset")}
          {renderNumber("0")}
          {renderOperator("=", "equals")}
          {renderOperator("/", "divide")}
        </tr>
      </tbody>
    </table>
  );
};

export default Keypad;
