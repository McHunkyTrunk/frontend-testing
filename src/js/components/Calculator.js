import React, { useState } from "react";

import { INITIAL_STATE } from "../config";
import { calculateNextState } from "../helpers/calculator";

import Display from "./Display";
import Keypad from "./Keypad";

const Calculator = () => {
  const [state, setState] = useState(INITIAL_STATE);

  return (
    <div className="Calculator">
      <Display mainValue={state.currentValue} secondaryValue={state.previousValue} />
      <Keypad onKey={key => setState(calculateNextState(state, key))} />
    </div>
  );
};

export default Calculator;
