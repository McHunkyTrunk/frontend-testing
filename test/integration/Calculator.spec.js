import chai, { expect } from "chai";
import chaiJestSnapshot from "chai-jest-snapshot";
import { JSDOM } from "jsdom";
import { after, before, describe, it } from "mocha";
import React from "react";
import ReactDOM from "react-dom";
import TestRenderer from "react-test-renderer";

import Calculator from "../../src/js/components/Calculator";
import Display from "../../src/js/components/Display";
import Keypad from "../../src/js/components/Keypad";

chai.use(chaiJestSnapshot);

before(function () {
  chaiJestSnapshot.resetSnapshotRegistry();
});

beforeEach(function () {
  chaiJestSnapshot.configureUsingMochaContext(this);
});

describe("Calculator component", function () {
  before(function () {
    global.window = new JSDOM(`
      <!doctype html>
      <html>
        <body>
          <div id="container"/></div>
        </body>
      </html>
    `).window;

    global.document = window.document;
  });

  after(function () {
    delete global.window;
    delete global.document;
  });

  it("should work (when tested with React DOM)", function () {
    ReactDOM.render(<Calculator />, document.getElementById("container"));

    const displayElement = document.querySelector(".Display");

    expect(displayElement.textContent).to.equal("0");

    const digit4Element = document.querySelector(".digit-4");
    const digit2Element = document.querySelector(".digit-2");
    const operatorMultiply = document.querySelector(".operator-multiply");
    const operatorEquals = document.querySelector(".operator-equals");

    digit4Element.click();
    digit2Element.click();
    operatorMultiply.click();
    digit2Element.click();
    operatorEquals.click();

    expect(displayElement.textContent).to.equal("84");
  });

  it("should work and match snapshots (when tested with React Test Reanderer)", function () {
    const component = TestRenderer.create(<Calculator />);

    expect(component.toJSON()).to.matchSnapshot();

    const KeypadElement = component.root.findByType(Keypad);

    // manually trigger some clicks on digits
    TestRenderer.act(() => KeypadElement.findByProps({"data-digit": "1"}).props.onClick());
    TestRenderer.act(() => KeypadElement.findByProps({"data-digit": "3"}).props.onClick());
    TestRenderer.act(() => KeypadElement.findByProps({"data-digit": "3"}).props.onClick());
    TestRenderer.act(() => KeypadElement.findByProps({"data-digit": "7"}).props.onClick());

    // re-rendering
    expect(component.toJSON()).to.matchSnapshot();

    // manually trigger a calculation
    TestRenderer.act(() => KeypadElement.findByProps({"data-operator": "/"}).props.onClick());
    TestRenderer.act(() => KeypadElement.findByProps({"data-digit": "7"}).props.onClick());
    TestRenderer.act(() => KeypadElement.findByProps({"data-operator": "="}).props.onClick());

    // re-rendering
    expect(component.toJSON()).to.matchSnapshot();

    const DisplayElement = component.root.findByType(Display);
    const valueElement = DisplayElement.findByProps({"data-testid": "Display__value"});

    expect(valueElement.children[0]).to.equal("191");
  });
});
