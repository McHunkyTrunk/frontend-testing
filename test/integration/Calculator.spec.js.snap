// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component should work and match snapshots (when tested with React Test Reanderer) 1`] = `
<div
  className="Calculator"
>
  <div
    className="Display"
    data-testid="Display"
  >
    <span
      className="main-value"
      data-testid="Display__value"
    >
      0
    </span>
    <span
      className="secondary-value"
    />
  </div>
  <table
    className="Keypad"
    data-testid="Keypad"
  >
    <tbody>
      <tr>
        <td
          className="digit digit-7"
          data-digit="7"
          data-testid="Keypad__digit-7"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className="digit digit-8"
          data-digit="8"
          data-testid="Keypad__digit-8"
          onClick={[Function]}
        >
          8
        </td>
        <td
          className="digit digit-9"
          data-digit="9"
          data-testid="Keypad__digit-9"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="operator operator-plus"
          data-operator="+"
          data-testid="Keypad__operator-plus"
          onClick={[Function]}
        >
          +
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-4"
          data-digit="4"
          data-testid="Keypad__digit-4"
          onClick={[Function]}
        >
          4
        </td>
        <td
          className="digit digit-5"
          data-digit="5"
          data-testid="Keypad__digit-5"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className="digit digit-6"
          data-digit="6"
          data-testid="Keypad__digit-6"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className="operator operator-subtract"
          data-operator="-"
          data-testid="Keypad__operator-subtract"
          onClick={[Function]}
        >
          -
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-1"
          data-digit="1"
          data-testid="Keypad__digit-1"
          onClick={[Function]}
        >
          1
        </td>
        <td
          className="digit digit-2"
          data-digit="2"
          data-testid="Keypad__digit-2"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className="digit digit-3"
          data-digit="3"
          data-testid="Keypad__digit-3"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className="operator operator-multiply"
          data-operator="*"
          data-testid="Keypad__operator-multiply"
          onClick={[Function]}
        >
          *
        </td>
      </tr>
      <tr>
        <td
          className="operator operator-reset"
          data-operator="C"
          data-testid="Keypad__operator-reset"
          onClick={[Function]}
        >
          C
        </td>
        <td
          className="digit digit-0"
          data-digit="0"
          data-testid="Keypad__digit-0"
          onClick={[Function]}
        >
          0
        </td>
        <td
          className="operator operator-equals"
          data-operator="="
          data-testid="Keypad__operator-equals"
          onClick={[Function]}
        >
          =
        </td>
        <td
          className="operator operator-divide"
          data-operator="/"
          data-testid="Keypad__operator-divide"
          onClick={[Function]}
        >
          /
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Calculator component should work and match snapshots (when tested with React Test Reanderer) 2`] = `
<div
  className="Calculator"
>
  <div
    className="Display"
    data-testid="Display"
  >
    <span
      className="main-value"
      data-testid="Display__value"
    >
      1337
    </span>
    <span
      className="secondary-value"
    >
      0
    </span>
  </div>
  <table
    className="Keypad"
    data-testid="Keypad"
  >
    <tbody>
      <tr>
        <td
          className="digit digit-7"
          data-digit="7"
          data-testid="Keypad__digit-7"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className="digit digit-8"
          data-digit="8"
          data-testid="Keypad__digit-8"
          onClick={[Function]}
        >
          8
        </td>
        <td
          className="digit digit-9"
          data-digit="9"
          data-testid="Keypad__digit-9"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="operator operator-plus"
          data-operator="+"
          data-testid="Keypad__operator-plus"
          onClick={[Function]}
        >
          +
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-4"
          data-digit="4"
          data-testid="Keypad__digit-4"
          onClick={[Function]}
        >
          4
        </td>
        <td
          className="digit digit-5"
          data-digit="5"
          data-testid="Keypad__digit-5"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className="digit digit-6"
          data-digit="6"
          data-testid="Keypad__digit-6"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className="operator operator-subtract"
          data-operator="-"
          data-testid="Keypad__operator-subtract"
          onClick={[Function]}
        >
          -
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-1"
          data-digit="1"
          data-testid="Keypad__digit-1"
          onClick={[Function]}
        >
          1
        </td>
        <td
          className="digit digit-2"
          data-digit="2"
          data-testid="Keypad__digit-2"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className="digit digit-3"
          data-digit="3"
          data-testid="Keypad__digit-3"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className="operator operator-multiply"
          data-operator="*"
          data-testid="Keypad__operator-multiply"
          onClick={[Function]}
        >
          *
        </td>
      </tr>
      <tr>
        <td
          className="operator operator-reset"
          data-operator="C"
          data-testid="Keypad__operator-reset"
          onClick={[Function]}
        >
          C
        </td>
        <td
          className="digit digit-0"
          data-digit="0"
          data-testid="Keypad__digit-0"
          onClick={[Function]}
        >
          0
        </td>
        <td
          className="operator operator-equals"
          data-operator="="
          data-testid="Keypad__operator-equals"
          onClick={[Function]}
        >
          =
        </td>
        <td
          className="operator operator-divide"
          data-operator="/"
          data-testid="Keypad__operator-divide"
          onClick={[Function]}
        >
          /
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Calculator component should work and match snapshots (when tested with React Test Reanderer) 3`] = `
<div
  className="Calculator"
>
  <div
    className="Display"
    data-testid="Display"
  >
    <span
      className="main-value"
      data-testid="Display__value"
    >
      191
    </span>
    <span
      className="secondary-value"
    />
  </div>
  <table
    className="Keypad"
    data-testid="Keypad"
  >
    <tbody>
      <tr>
        <td
          className="digit digit-7"
          data-digit="7"
          data-testid="Keypad__digit-7"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className="digit digit-8"
          data-digit="8"
          data-testid="Keypad__digit-8"
          onClick={[Function]}
        >
          8
        </td>
        <td
          className="digit digit-9"
          data-digit="9"
          data-testid="Keypad__digit-9"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="operator operator-plus"
          data-operator="+"
          data-testid="Keypad__operator-plus"
          onClick={[Function]}
        >
          +
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-4"
          data-digit="4"
          data-testid="Keypad__digit-4"
          onClick={[Function]}
        >
          4
        </td>
        <td
          className="digit digit-5"
          data-digit="5"
          data-testid="Keypad__digit-5"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className="digit digit-6"
          data-digit="6"
          data-testid="Keypad__digit-6"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className="operator operator-subtract"
          data-operator="-"
          data-testid="Keypad__operator-subtract"
          onClick={[Function]}
        >
          -
        </td>
      </tr>
      <tr>
        <td
          className="digit digit-1"
          data-digit="1"
          data-testid="Keypad__digit-1"
          onClick={[Function]}
        >
          1
        </td>
        <td
          className="digit digit-2"
          data-digit="2"
          data-testid="Keypad__digit-2"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className="digit digit-3"
          data-digit="3"
          data-testid="Keypad__digit-3"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className="operator operator-multiply"
          data-operator="*"
          data-testid="Keypad__operator-multiply"
          onClick={[Function]}
        >
          *
        </td>
      </tr>
      <tr>
        <td
          className="operator operator-reset"
          data-operator="C"
          data-testid="Keypad__operator-reset"
          onClick={[Function]}
        >
          C
        </td>
        <td
          className="digit digit-0"
          data-digit="0"
          data-testid="Keypad__digit-0"
          onClick={[Function]}
        >
          0
        </td>
        <td
          className="operator operator-equals"
          data-operator="="
          data-testid="Keypad__operator-equals"
          onClick={[Function]}
        >
          =
        </td>
        <td
          className="operator operator-divide"
          data-operator="/"
          data-testid="Keypad__operator-divide"
          onClick={[Function]}
        >
          /
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
